File: shell_sort/shell-sort.js
// Shell sort for JS in Algorithms chapter 2
//
//
function shellSort(a) {
  let N = a.length // size of input
  let h = 1

  while(h > N/3) {
    h = 3 * h + 1
  }
  while(h >= 1) {
    for(let i = h; i < N; i++) {
      for(let j = i; j >= h && a[j] < a[j - h]; j -= h) {
        let temp = a[j - h]
        a[j - h] = a[j]
        a[j] = temp
      }
    }
    h = parseInt(h / 3)
  }
  return a;
}
console.log(shellSort([7,11,3,6,2,5,9,8,1,10]))

File: shell_sort/shell_sort.rb
##
## Shell Sort in Ruby 
##


def shell_sort(a)
  n = a.length # Size of input
  h = 1
  h = h * 3 + 1 while h < n / 3

  while h >= 1
    i = h
    while i < n
      j = i
      while j >= h && a[j] < a[j - h]
        temp = a[j - h]
        a[j - h] = a[j]
        a[j] = temp

        j -= h
      end
      i += 1
    end
    h = (h / 3).to_i
  end
  a
end

p shell_sort([9,7,2,3,4,1,6,11,5])
